### ISSUE TITLE
[python] Dependency graph corner cases

### ISSUE THREAD

**Issue by tpetricek** [wrattler/wrattler#168]
Bug report from Ben Sach: 

> I think there might be some corner cases in the dependency graph that you’re missing just based on fiddling around (is there an easy way to see the dependency graph?).
> Maybe I’ve missed something though.
> 
> Two examples (all cells python).
> 
> ```
> # Cell 1
> df = pd.DataFrame({"x":[3,2,1]})
> # Cell 2
> df.sort_values(by='x', inplace=True)
> # Cell 3
> df2 = df
> ```
>
> Cell 3’s output is the same as Cell 1 (but should be 1,2,3). Merging cells 2 and 3 gives the correct output.

Hmm, I guess we assume that the data is not mutated - we do not re-export the imported cells automatically, which means that mutations remain cell-local. Workaround is to do `df = df`, but maybe we should just re-export everything (it's a bit unfortunate though) or maybe advise people to avoid mutation...

> Also…
> 
> ```
> # Cell 1
> df = pd.DataFrame({"x":[3,2,1]})
> # Cell 2
> df.x[0] = 100
> ```
>
> Gives name ‘df’ is not defined.

This is odd - @nbarlowATI do we correctly recognize `df.x[0]` as variable reference? That said, this will also suffer from the same problem as above.

**Comment by nbarlowATI** [comment]
I'll check about the df in df.x[0] being recognized - that's definitely a bug if not!

For mutations, I guess at the end of evaluating a cell we could check all the imports to see if they have changed from when we first retrieved them, and export them again if not.  The trouble is that we would only know this in 'eval', and I guess we need the information from 'exports'...

**Comment by tpetricek** [comment]
Yeah, exactly - we'd need to know this in `exports` and I don't think it's realistic to do code analysis to determine whether a data frame is mutated - so we can either not support that or re-export everything always. I'm somewhat more in favor of not supporting that for now.

**Comment by nbarlowATI** [comment]
I looked into the ```df.x[0] = 100``` case.  ```df``` was being recognized as a variable, but only added to "exports", while to access a specific element of it, it needed to be defined, and hence be in "imports".
Fix, which I think makes sense, is to treat this the same as df.some_method( )  which we had already come across, and so add the 'value' of 'ast.Subscript` nodes to the inputs_list.

This is implemented in ae32043113122b686f607c30c9207627861847cc in the bugfix/python_dependencies branch.
